File: src/main/java/cube8540/oauth/authentication/credentials/oauth/security/introspector/DefaultAccessTokenIntrospector.java
package cube8540.oauth.authentication.credentials.oauth.security.introspector;

import cube8540.oauth.authentication.credentials.oauth.security.OAuth2AccessTokenDetails;
import cube8540.oauth.authentication.credentials.oauth.security.OAuth2AccessTokenDetailsService;
import cube8540.oauth.authentication.credentials.oauth.security.provider.ClientCredentialsToken;
import cube8540.oauth.authentication.credentials.oauth.token.domain.OAuth2AccessTokenNotFoundException;
import lombok.Setter;
import org.springframework.security.authentication.AuthenticationProvider;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.oauth2.core.DefaultOAuth2AuthenticatedPrincipal;
import org.springframework.security.oauth2.core.OAuth2AuthenticatedPrincipal;
import org.springframework.security.oauth2.server.resource.introspection.OAuth2IntrospectionClaimNames;
import org.springframework.security.oauth2.server.resource.introspection.OAuth2IntrospectionException;
import org.springframework.security.oauth2.server.resource.introspection.OpaqueTokenIntrospector;

import java.util.Collection;
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;
import java.util.stream.Collectors;

public class DefaultAccessTokenIntrospector implements OpaqueTokenIntrospector {

    private final OAuth2AccessTokenDetailsService accessTokenService;

    private final AuthenticationProvider authenticationProvider;

    @Setter
    private String clientId;

    @Setter
    private String clientSecret;

    public DefaultAccessTokenIntrospector(OAuth2AccessTokenDetailsService accessTokenService, AuthenticationProvider authenticationProvider) {
        this.accessTokenService = accessTokenService;
        this.authenticationProvider = authenticationProvider;
    }

    @Override
    public OAuth2AuthenticatedPrincipal introspect(String token) {
        OAuth2AccessTokenDetails accessToken = readAccessToken(token);

        Authentication clientCredentials = clientAuthentication(clientId, clientSecret);
        if (!clientCredentials.getName().equals(accessToken.getClientId())) {
            throw new OAuth2IntrospectionException("client is different");
        }

        return new DefaultOAuth2AuthenticatedPrincipal(accessToken.getUsername(), convertClaims(accessToken), extractAuthorities(accessToken));
    }

    private OAuth2AccessTokenDetails readAccessToken(String token) {
        try {
            OAuth2AccessTokenDetails accessToken = accessTokenService.readAccessToken(token);
            if (accessToken.isExpired()) {
                throw new OAuth2IntrospectionException(token + " is not active");
            }
            return accessToken;
        } catch (OAuth2AccessTokenNotFoundException e) {
            throw new OAuth2IntrospectionException(token + " is not active", e);
        }
    }

    private Map<String, Object> convertClaims(OAuth2AccessTokenDetails accessToken) {
        Map<String, Object> claims = new HashMap<>();

        if (accessToken.getClientId() != null) {
            claims.put(OAuth2IntrospectionClaimNames.CLIENT_ID, accessToken.getClientId());
        }
        if (accessToken.getScopes() != null && !accessToken.getScopes().isEmpty()) {
            claims.put(OAuth2IntrospectionClaimNames.SCOPE, extractAuthorities(accessToken));
        }
        if (accessToken.getAdditionalInformation() != null) {
            accessToken.getAdditionalInformation().forEach(claims::put);
        }

        return claims;
    }

    private Collection<GrantedAuthority> extractAuthorities(OAuth2AccessTokenDetails accessToken) {
        if (accessToken.getScopes() != null && !accessToken.getScopes().isEmpty()) {
            return accessToken.getScopes().stream().map(SimpleGrantedAuthority::new).collect(Collectors.toList());
        } else {
            return Collections.emptySet();
        }
    }

    private Authentication clientAuthentication(String clientId, String clientSecret) {
        try {
            return authenticationProvider.authenticate(new ClientCredentialsToken(clientId, clientSecret));
        } catch (Exception e) {
            throw new OAuth2IntrospectionException("bad client credentials", e);
        }
    }
}

File: src/main/kotlin/cube8540/oauth/authentication/credentials/oauth/security/introspector/DefaultAccessTokenIntrospector.kt
package cube8540.oauth.authentication.credentials.oauth.security.introspector

import cube8540.oauth.authentication.credentials.oauth.security.OAuth2AccessTokenDetails
import cube8540.oauth.authentication.credentials.oauth.security.OAuth2AccessTokenDetailsService
import cube8540.oauth.authentication.credentials.oauth.security.provider.ClientCredentialsToken
import cube8540.oauth.authentication.credentials.oauth.token.domain.OAuth2AccessTokenNotFoundException
import org.springframework.security.authentication.AuthenticationProvider
import org.springframework.security.core.Authentication
import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.oauth2.core.DefaultOAuth2AuthenticatedPrincipal
import org.springframework.security.oauth2.core.OAuth2AuthenticatedPrincipal
import org.springframework.security.oauth2.server.resource.introspection.OAuth2IntrospectionClaimNames
import org.springframework.security.oauth2.server.resource.introspection.OAuth2IntrospectionException
import org.springframework.security.oauth2.server.resource.introspection.OpaqueTokenIntrospector
import java.util.*
import java.util.stream.Collectors
import kotlin.collections.HashMap

class DefaultAccessTokenIntrospector(private val accessTokenService: OAuth2AccessTokenDetailsService, private val authenticationProvider: AuthenticationProvider): OpaqueTokenIntrospector {

    var clientId: String? = null

    var clientSecret: String? = null

    override fun introspect(token: String): OAuth2AuthenticatedPrincipal {
        val accessToken = readAccessToken(token)

        val clientCredentials = clientAuthentication(clientId!!, clientSecret!!)
        if (clientCredentials.name != accessToken.clientId) {
            throw OAuth2IntrospectionException("Client is different")
        }
        return DefaultOAuth2AuthenticatedPrincipal(accessToken.username, convertClaims(accessToken), extractAuthorities(accessToken))
    }

    private fun readAccessToken(token: String): OAuth2AccessTokenDetails {
        try {
            val accessToken = accessTokenService.readAccessToken(token)
            if (accessToken.expired) {
                throw OAuth2IntrospectionException("$token is not active")
            }
            return accessToken
        } catch (e: OAuth2AccessTokenNotFoundException) {
            throw OAuth2IntrospectionException("$token is not active", e)
        }
    }

    private fun convertClaims(accessToken: OAuth2AccessTokenDetails): Map<String, *> {
        val claims = HashMap<String, Any?>()

        if (accessToken.clientId != null) {
            claims[OAuth2IntrospectionClaimNames.CLIENT_ID] = accessToken.clientId!!
        }
        if (accessToken.scopes != null && accessToken.scopes!!.isNotEmpty()) {
            claims[OAuth2IntrospectionClaimNames.SCOPE] = extractAuthorities(accessToken)
        }
        if (accessToken.additionalInformation != null) {
            accessToken.additionalInformation!!.forEach { (key, value) -> claims[key] = value }
        }

        return claims
    }

    private fun extractAuthorities(accessToken: OAuth2AccessTokenDetails): Collection<GrantedAuthority> {
        return if (accessToken.scopes != null && accessToken.scopes!!.isNotEmpty()) {
            accessToken.scopes!!.stream().map { grant -> SimpleGrantedAuthority(grant) }.collect(Collectors.toList())
        } else {
            Collections.emptySet()
        }
    }

    private fun clientAuthentication(clientId: String, clientSecret: String): Authentication {
        try {
            return authenticationProvider.authenticate(ClientCredentialsToken(clientId, clientSecret))
        } catch (e: Exception) {
            throw OAuth2IntrospectionException("Bad client credentials", e)
        }
    }
}